File: java/lib/src/main/java/com/svix/api/OperationalWebhookEndpoint.java
// this file is @generated
package com.svix.api;

import com.svix.SvixHttpClient;
import com.svix.Utils;
import com.svix.exceptions.ApiException;
import com.svix.models.ListResponseOperationalWebhookEndpointOut;
import com.svix.models.OperationalWebhookEndpointHeadersIn;
import com.svix.models.OperationalWebhookEndpointHeadersOut;
import com.svix.models.OperationalWebhookEndpointIn;
import com.svix.models.OperationalWebhookEndpointOut;
import com.svix.models.OperationalWebhookEndpointSecretIn;
import com.svix.models.OperationalWebhookEndpointSecretOut;
import com.svix.models.OperationalWebhookEndpointUpdate;

import okhttp3.Headers;
import okhttp3.HttpUrl;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class OperationalWebhookEndpoint {
    private final SvixHttpClient client;

    public OperationalWebhookEndpoint(SvixHttpClient client) {
        this.client = client;
    }

    /** List operational webhook endpoints. */
    public ListResponseOperationalWebhookEndpointOut list() throws IOException, ApiException {

        return this.list(new OperationalWebhookEndpointListOptions());
    }

    /** List operational webhook endpoints. */
    public ListResponseOperationalWebhookEndpointOut list(
            final OperationalWebhookEndpointListOptions options) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint");
        if (options.limit != null) {
            url.addQueryParameter("limit", Utils.serializeQueryParam(options.limit));
        }
        if (options.iterator != null) {
            url.addQueryParameter("iterator", options.iterator);
        }
        if (options.order != null) {
            url.addQueryParameter("order", Utils.serializeQueryParam(options.order));
        }
        return this.client.executeRequest(
                "GET", url.build(), null, null, ListResponseOperationalWebhookEndpointOut.class);
    }

    /** Create an operational webhook endpoint. */
    public OperationalWebhookEndpointOut create(
            final OperationalWebhookEndpointIn operationalWebhookEndpointIn)
            throws IOException, ApiException {
        return this.create(
                operationalWebhookEndpointIn, new OperationalWebhookEndpointCreateOptions());
    }

    /** Create an operational webhook endpoint. */
    public OperationalWebhookEndpointOut create(
            final OperationalWebhookEndpointIn operationalWebhookEndpointIn,
            final OperationalWebhookEndpointCreateOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint");
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        return this.client.executeRequest(
                "POST",
                url.build(),
                Headers.of(headers),
                operationalWebhookEndpointIn,
                OperationalWebhookEndpointOut.class);
    }

    /** Get an operational webhook endpoint. */
    public OperationalWebhookEndpointOut get(final String endpointId)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s", endpointId));
        return this.client.executeRequest(
                "GET", url.build(), null, null, OperationalWebhookEndpointOut.class);
    }

    /** Update an operational webhook endpoint. */
    public OperationalWebhookEndpointOut update(
            final String endpointId,
            final OperationalWebhookEndpointUpdate operationalWebhookEndpointUpdate)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s", endpointId));
        return this.client.executeRequest(
                "PUT",
                url.build(),
                null,
                operationalWebhookEndpointUpdate,
                OperationalWebhookEndpointOut.class);
    }

    /** Delete an operational webhook endpoint. */
    public void delete(final String endpointId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s", endpointId));
        this.client.executeRequest("DELETE", url.build(), null, null, null);
    }

    /** Get the additional headers to be sent with the operational webhook. */
    public OperationalWebhookEndpointHeadersOut getHeaders(final String endpointId)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s/headers",
                                        endpointId));
        return this.client.executeRequest(
                "GET", url.build(), null, null, OperationalWebhookEndpointHeadersOut.class);
    }

    /** Set the additional headers to be sent with the operational webhook. */
    public void updateHeaders(
            final String endpointId,
            final OperationalWebhookEndpointHeadersIn operationalWebhookEndpointHeadersIn)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s/headers",
                                        endpointId));
        this.client.executeRequest(
                "PUT", url.build(), null, operationalWebhookEndpointHeadersIn, null);
    }

    /**
     * Get an operational webhook endpoint's signing secret.
     *
     * <p>This is used to verify the authenticity of the webhook. For more information please refer
     * to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    public OperationalWebhookEndpointSecretOut getSecret(final String endpointId)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s/secret",
                                        endpointId));
        return this.client.executeRequest(
                "GET", url.build(), null, null, OperationalWebhookEndpointSecretOut.class);
    }

    /**
     * Rotates an operational webhook endpoint's signing secret.
     *
     * <p>The previous secret will remain valid for the next 24 hours.
     */
    public void rotateSecret(
            final String endpointId,
            final OperationalWebhookEndpointSecretIn operationalWebhookEndpointSecretIn)
            throws IOException, ApiException {
        this.rotateSecret(
                endpointId,
                operationalWebhookEndpointSecretIn,
                new OperationalWebhookEndpointRotateSecretOptions());
    }

    /**
     * Rotates an operational webhook endpoint's signing secret.
     *
     * <p>The previous secret will remain valid for the next 24 hours.
     */
    public void rotateSecret(
            final String endpointId,
            final OperationalWebhookEndpointSecretIn operationalWebhookEndpointSecretIn,
            final OperationalWebhookEndpointRotateSecretOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format(
                                        "/api/v1/operational-webhook/endpoint/%s/secret/rotate",
                                        endpointId));
        Map<String, String> headers = new HashMap<>();
        if (options.idempotencyKey != null) {
            headers.put("idempotency-key", options.idempotencyKey);
        }
        this.client.executeRequest(
                "POST", url.build(), Headers.of(headers), operationalWebhookEndpointSecretIn, null);
    }
}

File: kotlin/lib/src/main/kotlin/OperationalWebhookEndpoint.kt
// this file is @generated
package com.svix.kotlin

import com.svix.kotlin.models.ListResponseOperationalWebhookEndpointOut
import com.svix.kotlin.models.OperationalWebhookEndpointHeadersIn
import com.svix.kotlin.models.OperationalWebhookEndpointHeadersOut
import com.svix.kotlin.models.OperationalWebhookEndpointIn
import com.svix.kotlin.models.OperationalWebhookEndpointOut
import com.svix.kotlin.models.OperationalWebhookEndpointSecretIn
import com.svix.kotlin.models.OperationalWebhookEndpointSecretOut
import com.svix.kotlin.models.OperationalWebhookEndpointUpdate
import com.svix.kotlin.models.Ordering
import okhttp3.Headers

data class OperationalWebhookEndpointListOptions(
    /** Limit the number of returned items */
    val limit: ULong? = null,
    /** The iterator returned from a prior invocation */
    val iterator: String? = null,
    /** The sorting order of the returned items */
    val order: Ordering? = null,
)

data class OperationalWebhookEndpointCreateOptions(val idempotencyKey: String? = null)

data class OperationalWebhookEndpointRotateSecretOptions(val idempotencyKey: String? = null)

class OperationalWebhookEndpoint(private val client: SvixHttpClient) {
    /** List operational webhook endpoints. */
    suspend fun list(
        options: OperationalWebhookEndpointListOptions = OperationalWebhookEndpointListOptions()
    ): ListResponseOperationalWebhookEndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint")
        options.limit?.let { url.addQueryParameter("limit", serializeQueryParam(it)) }
        options.iterator?.let { url.addQueryParameter("iterator", it) }
        options.order?.let { url.addQueryParameter("order", serializeQueryParam(it)) }
        return client.executeRequest<Any, ListResponseOperationalWebhookEndpointOut>(
            "GET",
            url.build(),
        )
    }

    /** Create an operational webhook endpoint. */
    suspend fun create(
        operationalWebhookEndpointIn: OperationalWebhookEndpointIn,
        options: OperationalWebhookEndpointCreateOptions = OperationalWebhookEndpointCreateOptions(),
    ): OperationalWebhookEndpointOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<OperationalWebhookEndpointIn, OperationalWebhookEndpointOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = operationalWebhookEndpointIn,
        )
    }

    /** Get an operational webhook endpoint. */
    suspend fun get(endpointId: String): OperationalWebhookEndpointOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint/$endpointId")
        return client.executeRequest<Any, OperationalWebhookEndpointOut>("GET", url.build())
    }

    /** Update an operational webhook endpoint. */
    suspend fun update(
        endpointId: String,
        operationalWebhookEndpointUpdate: OperationalWebhookEndpointUpdate,
    ): OperationalWebhookEndpointOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint/$endpointId")

        return client.executeRequest<
            OperationalWebhookEndpointUpdate,
            OperationalWebhookEndpointOut,
        >(
            "PUT",
            url.build(),
            reqBody = operationalWebhookEndpointUpdate,
        )
    }

    /** Delete an operational webhook endpoint. */
    suspend fun delete(endpointId: String) {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint/$endpointId")
        client.executeRequest<Any, Boolean>("DELETE", url.build())
    }

    /** Get the additional headers to be sent with the operational webhook. */
    suspend fun getHeaders(endpointId: String): OperationalWebhookEndpointHeadersOut {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/headers")
        return client.executeRequest<Any, OperationalWebhookEndpointHeadersOut>("GET", url.build())
    }

    /** Set the additional headers to be sent with the operational webhook. */
    suspend fun updateHeaders(
        endpointId: String,
        operationalWebhookEndpointHeadersIn: OperationalWebhookEndpointHeadersIn,
    ) {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/headers")

        client.executeRequest<OperationalWebhookEndpointHeadersIn, Boolean>(
            "PUT",
            url.build(),
            reqBody = operationalWebhookEndpointHeadersIn,
        )
    }

    /**
     * Get an operational webhook endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook. For more information please refer to
     * [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    suspend fun getSecret(endpointId: String): OperationalWebhookEndpointSecretOut {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/secret")
        return client.executeRequest<Any, OperationalWebhookEndpointSecretOut>("GET", url.build())
    }

    /**
     * Rotates an operational webhook endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    suspend fun rotateSecret(
        endpointId: String,
        operationalWebhookEndpointSecretIn: OperationalWebhookEndpointSecretIn,
        options: OperationalWebhookEndpointRotateSecretOptions =
            OperationalWebhookEndpointRotateSecretOptions(),
    ) {
        val url =
            client
                .newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/secret/rotate")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        client.executeRequest<OperationalWebhookEndpointSecretIn, Boolean>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = operationalWebhookEndpointSecretIn,
        )
    }
}
